nvflare/private/fed/app/simulator/simulator_runner.py

```diff
@@ -571,9 +578,40 @@ def run_client_thread(self, num_of_threads, gpu, lock, rank, timeout=60):
                         self.end_run_clients.append(end_run_client)
 
                 client.simulate_running = False
+
+            if end_run_for_all:
+                self._end_run_clients(gpu, lock, num_of_threads, timeout)
         except Exception as e:
             self.logger.error(f"run_client_thread error: {secure_format_exception(e)}")
 
+    def _end_run_clients(self, gpu, lock, num_of_threads, timeout):
+        """After the WF reaches the END_RUN, each running thread will try to pick up one of the remaining client
+        which has not run the END_RUN yet, then execute the END_RUN handler, until all the clients have done so.
+        These client END_RUN event handler only execute when "end_run_for_all" has been set.
+
+        Multiple client running threads will try to pick up the client from the same clients pool.
+
+        """
+        # Each thread only stop picking up the NOT-DONE client until all clients have run the END_RUN event.
+        while len(self.end_run_clients) != len(self.federated_clients):
+            with lock:
+                end_run_client = self._pick_next_client()
+                self.end_run_clients.append(end_run_client.client_name)
+            if end_run_client:
+                self.do_one_task(
+                    end_run_client, num_of_threads, gpu, lock, timeout=timeout, task_name=RunnerTask.END_RUN
+                )
+                with lock:
```

Comment:
can't acquire lock inside with lock , will deadlock

Reply:
This is not acquire a lock inside another lock. They are separate.

Reply:
@YuanTingHsieh you mis-read the code. The first "with lock" only applies to line 598.